File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/Inventory/InventoryRepository.cs
namespace Assignment2Solution.Domain.Inventory;

/// <summary>
///     A very simple repository for managing the tea shop inventory.
/// </summary>
public sealed class InventoryRepository
{
    private readonly List<InventoryItem> _items;

    /// <summary>
    ///     Initializes a new instance of the <see cref="InventoryRepository" /> class with default items.
    /// </summary>
    public InventoryRepository()
    {
        _items = new List<InventoryItem>
        {
            new(Guid.NewGuid(), "Green Tea", 15.99m, 50, new StarRating(4)),
            new(Guid.NewGuid(), "Black Tea", 12.49m, 75, new StarRating(5)),
            new(Guid.NewGuid(), "Herbal Tea", 14.29m, 30, new StarRating(3)),
            new(Guid.NewGuid(), "Oolong Tea", 18.00m, 10, new StarRating(5)),
            new(Guid.NewGuid(), "Matcha", 29.99m, 0, new StarRating(4)),
            new(Guid.NewGuid(), "White Tea", 22.50m, 25, new StarRating(4)),
            new(Guid.NewGuid(), "Chai Tea", 10.99m, 60, new StarRating(3)),
            new(Guid.NewGuid(), "Earl Grey", 13.99m, 45, new StarRating(5)),
            new(Guid.NewGuid(), "Rooibos", 17.10m, 0, new StarRating(5)),
            new(Guid.NewGuid(), "Mint Tea", 11.89m, 80, new StarRating(1)),
            new(Guid.NewGuid(), "Jasmine Green", 16.75m, 35, new StarRating(4)),
            new(Guid.NewGuid(), "Genmaicha", 14.10m, 28, new StarRating(3)),
            new(Guid.NewGuid(), "Sencha", 19.25m, 40, new StarRating(4)),
            new(Guid.NewGuid(), "Darjeeling", 21.60m, 18, new StarRating(5)),
            new(Guid.NewGuid(), "Assam", 13.40m, 55, new StarRating(4)),
            new(Guid.NewGuid(), "Ceylon", 12.90m, 62, new StarRating(3)),
            new(Guid.NewGuid(), "Lapsang Souchong", 20.75m, 12, new StarRating(2)),
            new(Guid.NewGuid(), "Keemun", 17.35m, 22, new StarRating(4)),
            new(Guid.NewGuid(), "Pu-erh", 26.80m, 15, new StarRating(5)),
            new(Guid.NewGuid(), "Hojicha", 15.20m, 48, new StarRating(3)),
            new(Guid.NewGuid(), "Gyokuro", 32.50m, 8, new StarRating(5)),
            new(Guid.NewGuid(), "Bancha", 9.95m, 90, new StarRating(2)),
            new(Guid.NewGuid(), "Yerba Mate", 11.50m, 70, new StarRating(3)),
            new(Guid.NewGuid(), "Tulsi", 13.25m, 33, new StarRating(4)),
            new(Guid.NewGuid(), "Chamomile", 8.75m, 120, new StarRating(2)),
            new(Guid.NewGuid(), "Lavender", 9.60m, 44, new StarRating(2)),
            new(Guid.NewGuid(), "Lemongrass", 10.40m, 52, new StarRating(3)),
            new(Guid.NewGuid(), "Peppermint", 9.25m, 0, new StarRating(1)),
            new(Guid.NewGuid(), "Spearmint", 9.10m, 66, new StarRating(2)),
            new(Guid.NewGuid(), "Ginger Tea", 12.15m, 58, new StarRating(3)),
            new(Guid.NewGuid(), "Lemon Ginger", 11.80m, 47, new StarRating(3)),
            new(Guid.NewGuid(), "Turmeric Tea", 13.95m, 38, new StarRating(4)),
            new(Guid.NewGuid(), "Hibiscus", 10.25m, 41, new StarRating(2)),
            new(Guid.NewGuid(), "Rosehip", 10.55m, 29, new StarRating(3)),
            new(Guid.NewGuid(), "Berry Blend", 12.05m, 34, new StarRating(4)),
            new(Guid.NewGuid(), "Cinnamon Spice", 11.35m, 57, new StarRating(3)),
            new(Guid.NewGuid(), "Vanilla Chai", 14.85m, 26, new StarRating(4)),
            new(Guid.NewGuid(), "Masala Chai", 15.45m, 21, new StarRating(5)),
            new(Guid.NewGuid(), "Kashmiri Chai", 18.90m, 9, new StarRating(4)),
            new(Guid.NewGuid(), "London Fog", 13.70m, 31, new StarRating(3)),
            new(Guid.NewGuid(), "Breakfast Blend", 12.20m, 63, new StarRating(4)),
            new(Guid.NewGuid(), "English Breakfast", 11.95m, 77, new StarRating(4)),
            new(Guid.NewGuid(), "Irish Breakfast", 12.65m, 54, new StarRating(3)),
            new(Guid.NewGuid(), "Scottish Breakfast", 13.15m, 0, new StarRating(2)),
            new(Guid.NewGuid(), "Smoky Earl Grey", 14.55m, 24, new StarRating(5)),
            new(Guid.NewGuid(), "Orange Pekoe", 10.85m, 68, new StarRating(3)),
            new(Guid.NewGuid(), "Lemon Zest", 9.75m, 83, new StarRating(2)),
            new(Guid.NewGuid(), "Peach Oolong", 17.90m, 14, new StarRating(4)),
            new(Guid.NewGuid(), "Coconut Green", 16.40m, 0, new StarRating(3)),
            new(Guid.NewGuid(), "Caramel Rooibos", 18.35m, 19, new StarRating(4))
        };
    }

    /// <summary>
    ///     Gets all inventory items.
    /// </summary>
    /// <returns>A read-only list of all inventory items.</returns>
    public IReadOnlyList<InventoryItem> Get()
    {
        return _items.AsReadOnly();
    }

    /// <summary>
    ///     Updates the quantity of a specific inventory item.
    /// </summary>
    /// <param name="inventoryItemId">The ID of the item to update.</param>
    /// <param name="quantityChange">The amount to change the quantity by (positive or negative).</param>
    /// <exception cref="ArgumentException">Thrown when the item is not found.</exception>
    /// <exception cref="InvalidOperationException">Thrown when the resulting quantity would be negative.</exception>
    public void UpdateQuantity(Guid inventoryItemId, int quantityChange)
    {
        var index = _items.FindIndex(i => i.InventoryItemId == inventoryItemId);
        if (index == -1) throw new ArgumentException("Item not found", nameof(inventoryItemId));

        var current = _items[index];
        var newQuantity = current.Quantity + quantityChange;

        //invariant: quantity cannot be negative
        if (newQuantity < 0) throw new InvalidOperationException("Insufficient inventory.");

        // records allow 'with' expressions for non-destructive mutation
        _items[index] = current with { Quantity = newQuantity };
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventory/InventoryRepository.java
package assignment2solution.domain.inventory;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * A very simple repository for managing the tea shop inventory.
 */
public final class InventoryRepository {
    private final List<InventoryItem> items;

    /**
     * Initializes a new instance of the {@link InventoryRepository} class with default items.
     */
    public InventoryRepository() {
        items = new ArrayList<>(List.of(
            new InventoryItem(UUID.randomUUID(), "Green Tea", new BigDecimal("15.99"), 50, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Black Tea", new BigDecimal("12.49"), 75, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Herbal Tea", new BigDecimal("14.29"), 30, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Oolong Tea", new BigDecimal("18.00"), 10, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Matcha", new BigDecimal("29.99"), 0, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "White Tea", new BigDecimal("22.50"), 25, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Chai Tea", new BigDecimal("10.99"), 60, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Earl Grey", new BigDecimal("13.99"), 45, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Rooibos", new BigDecimal("17.10"), 0, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Mint Tea", new BigDecimal("11.89"), 80, new StarRating(1)),
            new InventoryItem(UUID.randomUUID(), "Jasmine Green", new BigDecimal("16.75"), 35, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Genmaicha", new BigDecimal("14.10"), 28, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Sencha", new BigDecimal("19.25"), 40, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Darjeeling", new BigDecimal("21.60"), 18, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Assam", new BigDecimal("13.40"), 55, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Ceylon", new BigDecimal("12.90"), 62, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Lapsang Souchong", new BigDecimal("20.75"), 12, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Keemun", new BigDecimal("17.35"), 22, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Pu-erh", new BigDecimal("26.80"), 15, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Hojicha", new BigDecimal("15.20"), 48, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Gyokuro", new BigDecimal("32.50"), 8, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Bancha", new BigDecimal("9.95"), 90, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Yerba Mate", new BigDecimal("11.50"), 70, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Tulsi", new BigDecimal("13.25"), 33, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Chamomile", new BigDecimal("8.75"), 120, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Lavender", new BigDecimal("9.60"), 44, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Lemongrass", new BigDecimal("10.40"), 52, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Peppermint", new BigDecimal("9.25"), 0, new StarRating(1)),
            new InventoryItem(UUID.randomUUID(), "Spearmint", new BigDecimal("9.10"), 66, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Ginger Tea", new BigDecimal("12.15"), 58, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Lemon Ginger", new BigDecimal("11.80"), 47, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Turmeric Tea", new BigDecimal("13.95"), 38, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Hibiscus", new BigDecimal("10.25"), 41, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Rosehip", new BigDecimal("10.55"), 29, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Berry Blend", new BigDecimal("12.05"), 34, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Cinnamon Spice", new BigDecimal("11.35"), 57, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Vanilla Chai", new BigDecimal("14.85"), 26, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Masala Chai", new BigDecimal("15.45"), 21, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Kashmiri Chai", new BigDecimal("18.90"), 9, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "London Fog", new BigDecimal("13.70"), 31, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Breakfast Blend", new BigDecimal("12.20"), 63, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "English Breakfast", new BigDecimal("11.95"), 77, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Irish Breakfast", new BigDecimal("12.65"), 54, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Scottish Breakfast", new BigDecimal("13.15"), 0, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Smoky Earl Grey", new BigDecimal("14.55"), 24, new StarRating(5)),
            new InventoryItem(UUID.randomUUID(), "Orange Pekoe", new BigDecimal("10.85"), 68, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Lemon Zest", new BigDecimal("9.75"), 83, new StarRating(2)),
            new InventoryItem(UUID.randomUUID(), "Peach Oolong", new BigDecimal("17.90"), 14, new StarRating(4)),
            new InventoryItem(UUID.randomUUID(), "Coconut Green", new BigDecimal("16.40"), 0, new StarRating(3)),
            new InventoryItem(UUID.randomUUID(), "Caramel Rooibos", new BigDecimal("18.35"), 19, new StarRating(4))
        ));
    }

    /**
     * Gets all inventory items.
     */
    public List<InventoryItem> get() {
        return Collections.unmodifiableList(items);
    }

    /**
     * Updates the quantity of a specific inventory item.
     *
     * @param inventoryItemId The ID of the item to update.
     * @param quantityChange  The amount to change the quantity by (positive or negative).
     * @throws IllegalArgumentException when the item is not found.
     * @throws IllegalStateException    when the resulting quantity would be negative.
     */
    public void updateQuantity(UUID inventoryItemId, int quantityChange) {
        int index = -1;
        for (int i = 0; i < items.size(); i++) {
            if (items.get(i).getInventoryItemId().equals(inventoryItemId)) {
                index = i;
                break;
            }
        }

        if (index == -1) {
            throw new IllegalArgumentException("Item not found");
        }

        var current = items.get(index);
        var newQuantity = current.getQuantity() + quantityChange;

        // invariant: quantity cannot be negative
        if (newQuantity < 0) {
            throw new IllegalStateException("Insufficient inventory.");
        }

        items.set(index, new InventoryItem(
            current.getInventoryItemId(),
            current.getName(),
            current.getPrice(),
            newQuantity,
            current.getStarRating()
        ));
    }
}
